### PR TITLE
feat(headercomponents): implement header left and right components

### PR DESCRIPTION
**Pull request by gcottrell1**
- If header left/right components are passed as props to the navigator, these will be rendered on all screens
- If header left/right is passed in when registering a screen, this will overwrite the default from the navigator and will receive the same props as the screen component

TODO: Update the peer dependency when https://github.com/ef-carbon/react-native-navigation/pull/35 is merged

### CODE REVIEW

File: lib/Backend.tsx [hunk 1]
```diff
@@ -120,8 +123,11 @@ export default class Backend implements IBackend {
     Object.defineProperties(this, descriptors);
```
Comment (mattyclarkson) [inline lib/Backend.tsx]: Add the `headerLeft`/`headerRight` descriptors here.

File: lib/Backend.tsx [hunk 2]
```diff
@@ -254,6 +263,16 @@ export default class Backend implements IBackend {
...
+          headerLeft: headerLeft ?
+            renderComponent(headerLeft, params)
+            : defaultHeaderLeft
+              ? renderComponent(defaultHeaderLeft)
+              : undefined,
+          headerRight: headerRight
+            ? renderComponent(headerRight, params)
+            : defaultHeaderRight
+              ? renderComponent(defaultHeaderRight)
+              : undefined,
```
Comment (mattyclarkson) [inline lib/Backend.tsx]: Might wanna make this into a function as it's repeated twice? Something like `renderHeaderComponent(headerLeft, defaultHeaderLeft, params)`?

File: lib/Backend.tsx [hunk 3]
```diff
@@ -77,6 +79,13 @@ function payload<P>(screen: ScreenComponentClass<P>, props: P): NavigationNaviga
   };
 }
 
+const renderHeaderComponent = (
+  override: Component<NavigationParams | undefined> | undefined,
+  defaultComponent: Component | undefined,
+  params: NavigationParams | undefined,
+): React.ReactElement<any> | undefined => // tslint:disable-line:no-any
```
Comment (mattyclarkson) [inline lib/Backend.tsx]: `Element` from `@ef-carbon/react-render-component` is the same as `React.ReactElement` but you don't need the `no-any` suppression.

Comment (gcottrell1) [reply lib/Backend.tsx]: Ye so I tried this first but typescript was complaining. I don't think I actually tried building though so maybe it was just a linting issue. I'll try this again

**Pull request merged**